doc/developer/design/20230119_ir_normalization.md

```diff
@@ -0,0 +1,147 @@
+# Intermediate Representation Normalization
+
+## Summary
+
+<!--
+// Brief, high-level overview. A few sentences long.
+// Be sure to capture the customer impact - framing this as a release note may be useful.
+-->
+
+Introduce normalizing transformations to some of HIR, MIR, and LIR that bring ASTs to a normal form with certain predictable properties.
+The intent is to introduce *predictability* to the representation of an AST, so that transformations, explanations, and other moments where Materialize and users interact with an expression AST do not have to worry or reason about as many equivalent alternatives.
+
+## Goals
+
+<!--
+// Enumerate the concrete goals that are in scope for the project.
+-->
+
+The main goal is to produce and maintain a `Normalize` transform that acts on a `MirRelationExpr` such that two structurally equivalent expressions become identical.
```

Comment:
nit: don't we want a transform that renders two semantically equivalent expressions structurally equivalent?

For example
```text
Filter #0 > 0
  Filter #0 < 5
    Get materialize.public.r
```
and
```text
Filter #0 > 0 and #0 < 5
    Get materialize.public.r
```
have different structure (that is, they are different `MirRelationExpr` trees) but they are extensionally equal (they will produce the same output for each input).